src/libsyntax/ext/base.rs

```diff
@@ -156,6 +164,173 @@ impl Into<Vec<Annotatable>> for Annotatable {
     }
 }
 
+pub trait ProcMacro {
+    fn expand<'cx>(&self,
+                   ecx: &'cx mut ExtCtxt,
+                   span: Span,
+                   ts: TokenStream)
+                   -> Box<MacResult+'cx>;
+}
+
+impl<F> ProcMacro for F
+    where F: Fn(TokenStream) -> TokenStream
+{
+    fn expand<'cx>(&self,
+                   ecx: &'cx mut ExtCtxt,
+                   span: Span,
+                   ts: TokenStream)
+                   -> Box<MacResult+'cx> {
+        let result = (*self)(ts);
+        // FIXME setup implicit context in TLS before calling self.
+        let parser = ecx.new_parser_from_tts(&result.to_tts());
+        Box::new(TokResult { parser: parser, span: span })
+    }
+}
+
+pub trait AttrProcMacro {
+    fn expand<'cx>(&self,
+                   ecx: &'cx mut ExtCtxt,
+                   span: Span,
+                   annotation: TokenStream,
+                   annotated: TokenStream)
+                   -> Box<MacResult+'cx>;
+}
+
+impl<F> AttrProcMacro for F
+    where F: Fn(TokenStream, TokenStream) -> TokenStream
+{
+    fn expand<'cx>(&self,
+                   ecx: &'cx mut ExtCtxt,
+                   span: Span,
+                   annotation: TokenStream,
+                   annotated: TokenStream)
+                   -> Box<MacResult+'cx> {
+        // FIXME setup implicit context in TLS before calling self.
+        let parser = ecx.new_parser_from_tts(&(*self)(annotation, annotated).to_tts());
+        Box::new(TokResult { parser: parser, span: span })
+    }
+}
+
+struct TokResult<'a> {
+    parser: Parser<'a>,
+    span: Span,
+}
+
+impl<'a> MacResult for TokResult<'a> {
+    fn make_items(mut self: Box<Self>) -> Option<SmallVector<P<ast::Item>>> {
+        if self.parser.sess.span_diagnostic.has_errors() {
+            return None;
+        }
+
+        let mut items = SmallVector::zero();
+        loop {
+            match self.parser.parse_item() {
+                Ok(Some(item)) => {
+                    // FIXME better span info.
+                    let mut item = item.unwrap();
+                    item.span = self.span;
```

Comment:
I regard this as something of a temporary solution - long term, proc macros should have expansion traces in the same way as macro rules today. The span we return here is going to be basically non-sensical - it will be the dummy span or a span pointing into some non-file part of the codemap, but pretty much something we shouldn't expose to the user. We could use the span of the proc macro, but given that will be the code *constructing* the code where the error occurs, it doesn't feel too useful.